File: src/test/java/com/contentful/java/cda/BaseTest.java
package com.contentful.java.cda;

import com.contentful.java.cda.lib.EnqueueResponseRule;
import com.contentful.java.cda.lib.TestCallback;
import com.contentful.java.cda.lib.TestResponse;

import org.apache.commons.io.FileUtils;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.List;
import java.util.logging.LogManager;

import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;

import static com.contentful.java.cda.Util.checkNotNull;
import static com.google.common.truth.Truth.assertThat;

public class BaseTest {
  public static final String DEFAULT_TOKEN = "test_token";
  public static final String DEFAULT_SPACE = "test_space";
  public static final String STAGING_ENVIRONMENT = "staging";

  CDAClient client;

  MockWebServer server;

  List<TestResponse> responseQueue;

  @Rule public EnqueueResponseRule enqueueResponse = new EnqueueResponseRule();

  @Before public void setUp() throws IOException {
    LogManager.getLogManager().reset();
    server = createServer();
    server.start();

    client = createClient();

    if (responseQueue != null) {
      for (TestResponse response : responseQueue) {
        enqueue(response);
      }
    }
  }

  @After public void tearDown() throws IOException {
    server.shutdown();
  }

  protected CDAClient createClient() {
    return createBuilder()
        .build();
  }

  protected CDAClient createPreviewClient() {
    return createBuilder()
        .preview()
        .setEndpoint(serverUrl())
        .build();
  }

  protected CDAClient.Builder createBuilder() {
    return CDAClient.builder()
        .setSpace(DEFAULT_SPACE)
        .setToken(DEFAULT_TOKEN)
        .setEndpoint(serverUrl());
  }

  protected String serverUrl() {
    return "http://" + server.getHostName() + ":" + server.getPort();
  }

  protected MockWebServer createServer() {
    return new MockWebServer();
  }

  protected void enqueue(TestResponse response) throws IOException {
    URL resource = getClass().getClassLoader().getResource(response.getFileName());
    checkNotNull(resource, "File not found: " + response.getFileName());
    server.enqueue(new MockResponse().setResponseCode(response.getCode())
        .setBody(FileUtils.readFileToString(new File(resource.getFile()), Charset.defaultCharset()))
        .setHeaders(response.headers()));
  }

  public BaseTest setResponseQueue(List<TestResponse> responseQueue) {
    this.responseQueue = responseQueue;
    return this;
  }

  protected <T extends CDAResource> T assertCallback(TestCallback<T> callback) {
    assertThat(callback.error()).isNull();
    assertThat(callback.result()).isNotNull();
    return callback.result();
  }
}

File: src/test/kotlin/com/contentful/java/cda/BaseTest.kt
/*
 * Copyright (C) 2014 Contentful GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.contentful.java.cda

import com.squareup.okhttp.mockwebserver.MockWebServer
import retrofit.RestAdapter
import org.junit.Before as before
import org.junit.After as after
import kotlin.test.assertNull
import kotlin.test.assertNotNull
import com.contentful.java.cda.lib.TestCallback
import com.squareup.okhttp.mockwebserver.MockResponse
import com.contentful.java.cda.lib.TestUtils

/**
 * BaseTest.
 */
open class BaseTest {
    var server: MockWebServer? = null
    var client: CDAClient? = null

    before fun setUp() {
        // MockWebServer
        server = MockWebServer()
        server!!.play()

        // Client
        client = CDAClient.Builder()
                .setAccessToken("token")
                .setSpaceKey("spaceid")
                .setEndpoint(getServerUrl())
                .setLogLevel(RestAdapter.LogLevel.FULL)
                .noSSL()
                .build()

        enqueue("space_fetch_response.json")
        client!!.assets().ensureSpace(true)
        server!!.takeRequest()
    }

    after fun tearDown() {
        server!!.shutdown()
    }

    fun <T> assertTestCallback(cb: CDACallback<T>): T {
        if (cb !is TestCallback) {
            throw IllegalArgumentException("callback should be an instance of TestCallback.")
        }
        cb.await()
        assertNull(cb.error)
        if (cb.allowEmpty) {
            return null
        }
        assertNotNull(cb.value)
        return cb.value!!
    }

    fun enqueue(fileName: String) {
        server!!.enqueue(MockResponse()
                .setResponseCode(200)
                .setBody(TestUtils.fileToString(fileName)))
    }

    fun getServerUrl(): String {
        val url = server!!.getUrl("/")
        return url.toString().substring(url.getProtocol().length() + 3)
    }
}
